### PR TITLE
[Open_Source_LLM] Add Dynamic List Functionalities

### PR DESCRIPTION
**Pull request by gjwoods**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by gjwoods** [comment]
Waiting on implementation to support this functionality

Build on the following PRs:
1. #825 - [Open_Source_LLM] Add Icon + MIR Telemetry Attribution + Fix Tests
2. #842 - [Open_Source_LLM] Add Endpoints Tests and switch Deployment setup

File: src/promptflow-tools/promptflow/tools/open_source_llm.py [hunk 1]
```diff
@@ -51,6 +51,92 @@ def wrapper(*args, **kwargs):
...
+                endpoint_key=ml_client.online_endpoints.get_keys(ep.name).primary_key)
+
+            list_of_endpoints.append(endpoint)
+
+            default_deployment_name = sorted(ep.traffic, key=lambda item: item[1])[0]
+
+            for d in ml_client.online_deployments.list(ep.name):
+                deployment = Deployment(
+                    deployment_name=d.name,
+                    model_family=get_model_type(d.model))
```
Comment (gjwoods) [inline src/promptflow-tools/promptflow/tools/open_source_llm.py:111]: If we can't parse a model, don't share the endpoint.

File: src/promptflow-tools/promptflow/tools/open_source_llm.py [hunk 2]
```diff
@@ -51,6 +51,100 @@ def wrapper(*args, **kwargs):
...
+class EndpointsContainer:
+    '''We're probably going to need to cache this somehow'''
+    def __init__(self) -> None:
+        self.__endpoints_and_deployments = None
+
+    def get_ml_client(self) -> MLClient:
+        from azure.identity import DefaultAzureCredential
+        credential = DefaultAzureCredential(exclude_interactive_browser_credential=False)
+        try:
+            return MLClient(
```
Comment (chjinche) [inline src/promptflow-tools/promptflow/tools/open_source_llm.py:83]: early fyi, may not need to set os env, in local user just need to az login and configure azure ws info.

System will append ws triple to list function input params if the function signature has 'subscription_id','resource_group_name','workspace_name' or `**kwargs`. More references:
https://github.com/microsoft/promptflow/blob/a589f8281aa5fd22379c5bff74c968af0bf4b66a/src/promptflow/promptflow/_core/tools_manager.py#L181

**Review by prakharg-msft — APPROVED** [review]

File: src/promptflow-tools/promptflow/tools/yamls/open_source_llm.yaml [hunk 3]
```diff
@@ -8,17 +8,27 @@ promptflow.tools.open_source_llm.OpenSourceLLM.call:
...
     connection:
       default: null
       type:
       - CustomConnection
     deployment_name:
-      default: null
       type:
       - string
+      dynamic_list:
+        func_path: promptflow.tools.open_source_llm.EndpointsContainer.list_deployment_names
```
Comment (chjinche) [inline src/promptflow-tools/promptflow/tools/yamls/open_source_llm.yaml:25]: In first version of dynamic list support (which we are releasing this week), class/instance method is not supported. Please kindly change to use function. See this as an example https://github.com/microsoft/promptflow/blob/21ef07ea35fb101b11b845c22f4cbeb096fb5760/examples/tools/tool-package-quickstart/my_tool_package/tools/tool_with_dynamic_list_input.py#L5

**Pull request closed without merging**